File: sdks/go/insforge/storage.go
package insforge

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// Storage provides file storage operations.
type Storage struct {
	http *httpClient
}

func newStorage(h *httpClient) *Storage { return &Storage{http: h} }

// From returns a StorageBucket for the given bucket name.
func (s *Storage) From(bucketName string) *StorageBucket {
	return &StorageBucket{http: s.http, bucket: bucketName}
}

// StorageBucket wraps operations for a specific bucket.
type StorageBucket struct {
	http   *httpClient
	bucket string
}

func (b *StorageBucket) base() string {
	return "/api/storage/buckets/" + b.bucket
}

// GetPublicURL returns the public URL for an object in the bucket.
func (b *StorageBucket) GetPublicURL(path string) string {
	clean := strings.TrimLeft(path, "/")
	return fmt.Sprintf("%s%s/objects/%s", b.http.baseURL, b.base(), clean)
}

// Upload uploads raw bytes to the given path in the bucket.
func (b *StorageBucket) Upload(ctx context.Context, path string, data []byte, contentType string) Result[interface{}] {
	if contentType == "" {
		contentType = "application/octet-stream"
	}
	clean := strings.TrimLeft(path, "/")

	// Request upload strategy
	stratRaw, err := b.http.post(ctx, b.base()+"/upload-strategy", map[string]interface{}{
		"filename": clean, "contentType": contentType, "size": len(data),
	}, nil)
	if err != nil {
		return fail[interface{}](err)
	}

	strategy, _ := stratRaw.(map[string]interface{})
	method, _ := strategy["method"].(string)
	uploadURL, _ := strategy["uploadUrl"].(string)
	key, _ := strategy["key"].(string)
	if key == "" {
		key = clean
	}

	if method == "presigned" && uploadURL != "" {
		// S3 presigned POST: multipart/form-data with strategy fields + file content
		var body bytes.Buffer
		mw := multipart.NewWriter(&body)
		// Write all signed fields first (policy, signature, key, etc.)
		if fields, ok := strategy["fields"].(map[string]interface{}); ok {
			for k, v := range fields {
				_ = mw.WriteField(k, fmt.Sprintf("%v", v))
			}
		}
		// File content must be last
		fw, err := mw.CreateFormFile("file", key)
		if err != nil {
			return fail[interface{}](err)
		}
		if _, err = fw.Write(data); err != nil {
			return fail[interface{}](err)
		}
		mw.Close()

		req, err := http.NewRequestWithContext(ctx, http.MethodPost, uploadURL, &body)
		if err != nil {
			return fail[interface{}](err)
		}
		req.Header.Set("Content-Type", mw.FormDataContentType())
		resp, err := b.http.client.Do(req)
		if err != nil {
			return fail[interface{}](err)
		}
		defer resp.Body.Close()
		if resp.StatusCode >= 400 {
			return fail[interface{}](&InsForgeError{StatusCode: resp.StatusCode, Message: "presigned upload failed"})
		}
		// Confirm if required
		confirmReq, _ := strategy["confirmRequired"].(bool)
		if confirmReq {
			confirmURL, _ := strategy["confirmUrl"].(string)
			// confirmUrl may be a relative path or full URL; strip base URL if present
			confirmPath := strings.Replace(confirmURL, b.http.baseURL, "", 1)
			raw, err := b.http.post(ctx, confirmPath, map[string]interface{}{"size": len(data), "contentType": contentType}, nil)
			if err != nil {
				return fail[interface{}](err)
			}
			return ok[interface{}](raw)
		}
		return ok[interface{}](map[string]interface{}{"key": key})
	}

	// Direct upload: use the uploadUrl from strategy (has the URL-encoded key)
	directPath := strings.Replace(uploadURL, b.http.baseURL, "", 1)
	if directPath == "" {
		directPath = b.base() + "/objects/" + key
	}
	raw, err := b.http.uploadRaw(ctx, directPath, data, contentType)
	if err != nil {
		return fail[interface{}](err)
	}
	return ok[interface{}](raw)
}

// UploadReader uploads from an io.Reader.
func (b *StorageBucket) UploadReader(ctx context.Context, path string, r io.Reader, contentType string) Result[interface{}] {
	data, err := io.ReadAll(r)
	if err != nil {
		return fail[interface{}](&InsForgeError{Message: "failed to read upload data: " + err.Error()})
	}
	return b.Upload(ctx, path, data, contentType)
}

// Download downloads an object and returns its raw bytes.
func (b *StorageBucket) Download(ctx context.Context, path string) Result[[]byte] {
	clean := strings.TrimLeft(path, "/")

	// Get download strategy
	stratRaw, err := b.http.post(ctx, b.base()+"/objects/"+clean+"/download-strategy",
		map[string]interface{}{"expiresIn": 3600}, nil)
	if err != nil {
		// Fallback: direct download
		data, derr := b.http.downloadBytes(ctx, b.base()+"/objects/"+clean)
		if derr != nil {
			return fail[[]byte](err)
		}
		return ok(data)
	}

	strategy, _ := stratRaw.(map[string]interface{})
	method, _ := strategy["method"].(string)
	downloadURL, _ := strategy["url"].(string)

	if downloadURL == "" {
		data, err := b.http.downloadBytes(ctx, b.base()+"/objects/"+clean)
		if err != nil {
			return fail[[]byte](err)
		}
		return ok(data)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, downloadURL, nil)
	if err != nil {
		return fail[[]byte](err)
	}
	// For direct downloads, include auth headers
	if method == "direct" {
		if token := b.http.authToken(); token != "" {
			req.Header.Set("Authorization", "Bearer "+token)
		}
		for k, v := range b.http.headers {
			req.Header.Set(k, v)
		}
	}
	resp, err := b.http.client.Do(req)
	if err != nil {
		return fail[[]byte](err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return fail[[]byte](&InsForgeError{StatusCode: resp.StatusCode, Message: "download failed"})
	}
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return fail[[]byte](err)
	}
	return ok(data)
}

// List lists objects in the bucket.
func (b *StorageBucket) List(ctx context.Context, opts *ListOptions) Result[interface{}] {
	params := url.Values{}
	params.Set("limit", "100")
	params.Set("offset", "0")
	if opts != nil {
		if opts.Prefix != "" {
			params.Set("prefix", opts.Prefix)
		}
		if opts.Search != "" {
			params.Set("search", opts.Search)
		}
		if opts.Limit > 0 {
			params.Set("limit", strconv.Itoa(opts.Limit))
		}
		if opts.Offset > 0 {
			params.Set("offset", strconv.Itoa(opts.Offset))
		}
	}
	raw, err := b.http.get(ctx, b.base()+"/objects", params, nil)
	if err != nil {
		return fail[interface{}](err)
	}
	return ok[interface{}](raw)
}

// Remove deletes an object from the bucket.
func (b *StorageBucket) Remove(ctx context.Context, path string) Result[interface{}] {
	clean := strings.TrimLeft(path, "/")
	raw, err := b.http.delete(ctx, b.base()+"/objects/"+clean, nil, nil)
	if err != nil {
		return fail[interface{}](err)
	}
	return ok[interface{}](raw)
}

// ListOptions controls list pagination and filtering.
type ListOptions struct {
	Prefix string
	Search string
	Limit  int
	Offset int
}

File: sdks/python/insforge/storage.py
"""
Storage module for the InsForge Python SDK.

Usage:
    result = client.storage.from_("images").upload("posts/cover.jpg", file_bytes, "image/jpeg")
    result = client.storage.from_("uploads").upload_auto(file_bytes, "image/png")
    data   = client.storage.from_("images").download("posts/cover.jpg")
    result = client.storage.from_("images").remove("posts/cover.jpg")
"""

from __future__ import annotations

import io
import mimetypes
import uuid
from typing import Any, BinaryIO

from requests import RequestException

from .http import InsForgeError


class StorageBucket:
    """Scoped operations on a single bucket."""

    def __init__(self, http: Any, bucket_name: str) -> None:
        self._http = http
        self._bucket = bucket_name

    # ------------------------------------------------------------------
    # Upload
    # ------------------------------------------------------------------

    def upload(
        self,
        path: str,
        file: bytes | BinaryIO,
        content_type: str | None = None,
    ) -> dict[str, Any]:
        """
        Upload a file at the given path/key.

        Uses the new upload-strategy API automatically:
        - For local storage → direct PUT
        - For S3 → presigned POST then confirm

        Args:
            path: Object key (e.g. "users/avatar.jpg").
            file: File bytes or a file-like object.
            content_type: MIME type. Auto-detected from path if omitted.

        Returns:
            Dict with keys: bucket, key, size, mime_type, uploaded_at, url.

        Raises:
            InsForgeError: On API error.
        """
        if isinstance(file, (bytes, bytearray)):
            content = bytes(file)
        else:
            content = file.read()

        if content_type is None:
            content_type, _ = mimetypes.guess_type(path)
            content_type = content_type or "application/octet-stream"

        filename = path
        size = len(content)

        # Step 1: get upload strategy
        strategy = self._http.post(
            f"/api/storage/buckets/{self._bucket}/upload-strategy",
            data={"filename": filename, "contentType": content_type, "size": size},
        )

        method = strategy.get("method", "direct")
        upload_url = strategy["uploadUrl"]
        key = strategy["key"]
        confirm_required = strategy.get("confirmRequired", False)

        if method == "presigned":
            # S3 presigned POST
            fields = strategy.get("fields", {})
            multipart_data = {k: (None, v) for k, v in fields.items()}
            multipart_data["file"] = (filename, io.BytesIO(content), content_type)
            self._http.post_external(upload_url, files=multipart_data)

            if confirm_required:
                # Server explicitly requires a confirm round-trip
                confirm_url = strategy["confirmUrl"]
                return self._http.post(
                    confirm_url,
                    data={"size": size, "contentType": content_type},
                )
            elif "confirmUrl" in strategy:
                # confirmUrl is present but confirmRequired is False — still call it
                return self._http.post(
                    strategy["confirmUrl"],
                    data={"size": size, "contentType": content_type},
                )
            else:
                # No confirmUrl — use the same metadata endpoint as the PUT branch
                return self._http.post(
                    f"/api/storage/buckets/{self._bucket}/objects/{key}/confirm-upload",
                    data={"size": size, "contentType": content_type},
                )
        else:
            # Direct upload via PUT
            self._http.put_external(
                self._http._url(upload_url),
                content=content,
                content_type=content_type,
            )
            # Fetch object metadata
            return self._http.post(
                f"/api/storage/buckets/{self._bucket}/objects/{key}/confirm-upload",
                data={"size": size, "contentType": content_type},
            )

    def upload_auto(
        self,
        file: bytes | BinaryIO,
        content_type: str | None = None,
        original_filename: str | None = None,
    ) -> dict[str, Any]:
        """
        Upload a file with an auto-generated unique key.

        Args:
            file: File bytes or a file-like object.
            content_type: MIME type.
            original_filename: Hint for extension detection.

        Returns:
            Dict with keys: bucket, key, size, mime_type, uploaded_at, url.
        """
        filename = original_filename or f"upload-{uuid.uuid4().hex}"
        return self.upload(filename, file, content_type)

    # ------------------------------------------------------------------
    # Download
    # ------------------------------------------------------------------

    def download(self, path: str) -> dict[str, Any]:
        """
        Download a file as bytes.

        Uses the download-strategy API:
        - Local storage → direct GET
        - S3 → fetches from presigned URL

        Args:
            path: Object key.

        Returns:
            Dict with keys: data (bytes), error.
        """
        try:
            strategy = self._http.post(
                f"/api/storage/buckets/{self._bucket}/objects/{path}/download-strategy",
                data={},
            )
            url = strategy.get("url", "")
            if url.startswith("http"):
                # External URL (S3 presigned) — use the SDK's configured session
                r = self._http._session.get(url, timeout=self._http.timeout)
                r.raise_for_status()
                return {"data": r.content, "error": None}
            else:
                # Local storage — relative URL
                content = self._http.get_raw(url)
                return {"data": content, "error": None}
        except (RequestException, InsForgeError) as exc:
            return {"data": None, "error": exc}

    # ------------------------------------------------------------------
    # Remove
    # ------------------------------------------------------------------

    def remove(self, path: str) -> dict[str, Any]:
        """
        Delete a file from the bucket.

        Args:
            path: Object key.

        Returns:
            Dict with keys: data (message), error.
        """
        try:
            result = self._http.delete(
                f"/api/storage/buckets/{self._bucket}/objects/{path}"
            )
            return {"data": result, "error": None}
        except Exception as exc:
            return {"data": None, "error": exc}

    # ------------------------------------------------------------------
    # List
    # ------------------------------------------------------------------

    def list(
        self,
        *,
        prefix: str | None = None,
        search: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List objects in the bucket.

        Args:
            prefix: Filter by key prefix.
            search: Search by partial key match.
            limit: Max objects to return (1-1000).
            offset: Number of objects to skip.

        Returns:
            Dict with keys: data (list of objects), pagination, error.
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if prefix:
            params["prefix"] = prefix
        if search:
            params["search"] = search
        try:
            result = self._http.get(
                f"/api/storage/buckets/{self._bucket}/objects", params=params
            )
            return {"data": result.get("data", []), "pagination": result.get("pagination"), "error": None}
        except Exception as exc:
            return {"data": None, "pagination": None, "error": exc}


class StorageClient:
    """Entry point for storage operations."""

    def __init__(self, http: Any) -> None:
        self._http = http

    def from_(self, bucket_name: str) -> StorageBucket:
        """
        Get a bucket instance for file operations.

        Args:
            bucket_name: Name of the storage bucket.

        Returns:
            StorageBucket instance.

        Example:
            bucket = client.storage.from_("images")
            result = bucket.upload("cover.jpg", data, "image/jpeg")
        """
        return StorageBucket(self._http, bucket_name)
